### PR TITLE
update uptimerobot

### PR DESCRIPTION
**Pull request by Ninjiner**
##### SUMMARY
Although this code needs improvement, it makes it possible to add and remove monitors from uptimerobot. Also it interacts with the api v2 of uptimerobot. These three essential functionalities were missing before and I think they are the main reason you would use this module.

This would resolve this two issues:
* Fixes https://github.com/ansible/ansible/issues/36984
* Fixes https://github.com/ansible/ansible/issues/56994

##### ISSUE TYPE
* Feature Pull Request

##### COMPONENT NAME
uptimerobot

##### ADDITIONAL INFORMATION
I originally made [this](https://github.com/ansible/ansible/pull/63843) PR in the ansible/ansible repo. Because I would love to see the changes beeing merged, I am transfering this PR to this repo.

### CODE REVIEW

File: plugins/modules/monitoring/uptimerobot.py [hunk 1]
```diff
@@ -7,139 +7,183 @@
...
         description:
             - Define whether or not the monitor should be running or paused.
         required: true
-        choices: [ "started", "paused" ]
+        choices: [ "started", "paused", "created", "absent", "present" ]
+    url:
+        description:
+            - URL to be checked.
+        required: true
+        version_added: '2.10'
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:35]:
```suggestion
        version_added: '1.2.0'
```
Same for the others.

File: plugins/modules/monitoring/uptimerobot.py [hunk 2]
```diff
@@ -7,139 +7,183 @@
...
             - ID of the monitor to check.
-        required: true
+        required: false
     apikey:
         description:
             - Uptime Robot API key.
         required: true
 notes:
-    - Support for adding and removing monitors and alert contacts has not yet been implemented.
+    - Support for further functionalities, which the api provides has not yet been implemented.
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:55]:
```suggestion
    - Support for further functionalities, which the API provides has not yet been implemented.
```

File: plugins/modules/monitoring/uptimerobot.py [hunk 3]
```diff
@@ -7,139 +7,183 @@
...
+    - created and absent are added in version 2.10
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:56]:
```suggestion
```
This should be mentioned in the `state` documentation.

File: plugins/modules/monitoring/uptimerobot.py [hunk 4]
```diff
@@ -7,139 +7,183 @@
...
 '''
 
 EXAMPLES = '''
-- name: Pause the monitor with an ID of 12345
-  community.general.uptimerobot:
-    monitorid: 12345
+# Pause the monitor with an ID of 12345.
+- uptimerobot:
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:61]: Use `name:` and FQCN.

File: plugins/modules/monitoring/uptimerobot.py [hunk 5]
```diff
@@ -7,139 +7,183 @@
...
     module = AnsibleModule(
         argument_spec=dict(
-            state=dict(required=True, choices=['started', 'paused']),
+            name=dict(required=True),
+            url=dict(required=True),
+            check_type=dict(required=False, choices=['http', 'ping']),
+            state=dict(required=True, choices=['started', 'paused', 'absent', 'created', 'present']),
             apikey=dict(required=True, no_log=True),
-            monitorid=dict(required=True)
+            monitorid=dict(required=False)
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:163]: Is this no longer required, at least for `state='started'` or `state='paused'`? I don't see you ensuring this anywhere.

**Review by felixfontein — COMMENTED** [review]
Some first comments.

File: plugins/modules/monitoring/uptimerobot.py (hunk 1 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Fixed in: https://github.com/ansible-collections/community.general/pull/961/commits/12ad8ea0388fe96e3606eb99335cec60233185f3

File: plugins/modules/monitoring/uptimerobot.py (hunk 3 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/863c2e89453a20fcce643f8492160a4c4bba2b1d

File: plugins/modules/monitoring/uptimerobot.py (hunk 2 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Fixed in: https://github.com/ansible-collections/community.general/pull/961/commits/863c2e89453a20fcce643f8492160a4c4bba2b1d

File: plugins/modules/monitoring/uptimerobot.py [hunk 6]
```diff
@@ -7,139 +7,183 @@
 __metaclass__ = type
 
 
-DOCUMENTATION = '''
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
[the remark this answers is not in the document]
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Removed in: https://github.com/ansible-collections/community.general/pull/961/commits/2212a1ee602d5e151782ee646f4f49d852d39f95

File: plugins/modules/monitoring/uptimerobot.py [hunk 7]
```diff
@@ -8,138 +8,178 @@
...
-short_description: Pause and start Uptime Robot monitoring
+short_description: Create, remove, pause and start monitors in Uptimerobot
 description:
-    - This module will let you start and pause Uptime Robot Monitoring
-author: "Nate Kingsley (@nate-kingsley)"
+    - This module lets you create, remove, pause and start monitors in Uptimerobot
+author:
+- "Nate Kingsley (@nate-kingsley)"
+- "Ninjiner (@Ninjiner)"
+version_added: "1.9"
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:18]:
```suggestion
```

File: plugins/modules/monitoring/uptimerobot.py [hunk 8]
```diff
@@ -8,138 +8,178 @@
...
 
 DOCUMENTATION = '''
-
 module: uptimerobot
-short_description: Pause and start Uptime Robot monitoring
+short_description: Create, remove, pause and start monitors in Uptimerobot
 description:
-    - This module will let you start and pause Uptime Robot Monitoring
-author: "Nate Kingsley (@nate-kingsley)"
+    - This module lets you create, remove, pause and start monitors in Uptimerobot
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:14]:
```suggestion
    - This module lets you create, remove, pause and start monitors in Uptimerobot.
```

File: plugins/modules/monitoring/uptimerobot.py [hunk 9]
```diff
@@ -8,138 +8,178 @@
 
 
 DOCUMENTATION = '''
-
 module: uptimerobot
-short_description: Pause and start Uptime Robot monitoring
+short_description: Create, remove, pause and start monitors in Uptimerobot
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:12]: Is `Uptimerobot` the official spelling of the project? Or why did you change `Uptime Robot` to `Uptimerobot`?

File: plugins/modules/monitoring/uptimerobot.py [hunk 10]
```diff
@@ -8,138 +8,178 @@
...
+- "Nate Kingsley (@nate-kingsley)"
+- "Ninjiner (@Ninjiner)"
+version_added: "1.9"
 requirements:
     - Valid Uptime Robot API Key
 options:
     state:
         description:
             - Define whether or not the monitor should be running or paused.
+            - created and absent are added in version 1.2.0
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:25]:
```suggestion
            - C(created) and C(absent) were added in community.general 1.3.0.
```

File: plugins/modules/monitoring/uptimerobot.py [hunk 11]
```diff
@@ -7,139 +7,183 @@
...
-short_description: Pause and start Uptime Robot monitoring
+short_description: Create, remove, pause and start monitors in Uptimerobot
 description:
-    - This module will let you start and pause Uptime Robot Monitoring
-author: "Nate Kingsley (@nate-kingsley)"
+    - This module lets you create, remove, pause and start monitors in Uptimerobot
+author:
+- "Nate Kingsley (@nate-kingsley)"
+- "Ninjiner (@Ninjiner)"
+version_added: "1.9"
```
[the remark this answers is not in the document]
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/1b23056c9a8000f9bf8dc3063ca137bd9d179ccf

File: plugins/modules/monitoring/uptimerobot.py (hunk 7 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/1b23056c9a8000f9bf8dc3063ca137bd9d179ccf

File: plugins/modules/monitoring/uptimerobot.py (hunk 9 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/5d5c6bcc163c83624ae1294289a67fdc47fb9029

File: plugins/modules/monitoring/uptimerobot.py (hunk 10 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in https://github.com/ansible-collections/community.general/pull/961/commits/5d5c6bcc163c83624ae1294289a67fdc47fb9029

File: plugins/modules/monitoring/uptimerobot.py (hunk 8 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/5d5c6bcc163c83624ae1294289a67fdc47fb9029

File: plugins/modules/monitoring/uptimerobot.py (hunk 10 above)
Comment (felixfontein) [reply plugins/modules/monitoring/uptimerobot.py]: You didn't use `C(...)` and also the wrong case for the options.

Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in https://github.com/ansible-collections/community.general/pull/961/commits/dab1a424d7eb88d6ab69406896be8141a8e78f77

File: plugins/modules/monitoring/uptimerobot.py [hunk 12]
```diff
@@ -8,138 +8,177 @@
...
             - Define whether or not the monitor should be running or paused.
+            - C(created) and C(absent) were added in community.general 1.3.0.
+        required: true
+        choices: [ "started", "paused", "created", "absent", "present" ]
+    url:
+        description:
+            - URL to be checked.
         required: true
-        choices: [ "started", "paused" ]
+        version_added: '1.2.0'
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:31]:
```suggestion
        version_added: '1.3.0'
```
(Same for the others below.)

Comment (felixfontein) [reply plugins/modules/monitoring/uptimerobot.py:31]: Please add `type: ...` for the new options. That they are not there for the old ones is a sanity error that is ignored for this module, but is something that should also be changed eventually.

Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed the version strings: https://github.com/ansible-collections/community.general/pull/961/commits/10ebea24634059a54acd8bdddc4bfba560446209

Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Added data types: https://github.com/ansible-collections/community.general/pull/961/commits/d4d073d4be2a5145d7570f58812eb31a35242bf7

File: plugins/modules/monitoring/uptimerobot.py [hunk 13]
```diff
@@ -8,138 +8,181 @@
...
+author:
+- "Nate Kingsley (@nate-kingsley)"
+- "Ninjiner (@Ninjiner)"
 requirements:
     - Valid Uptime Robot API Key
 options:
     state:
         description:
             - Define whether or not the monitor should be running or paused.
+            - C(created) and C(absent) were added in community.general 1.3.0.
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:24]: `present` wasn't there before, either.

File: plugins/modules/monitoring/uptimerobot.py [hunk 14]
```diff
@@ -8,138 +8,181 @@
...
+        required: true
+        version_added: '1.3.0'
+    check_type:
+        description:
+            - The kind of check, that will be performed on the url.
+        choices: [ "http", "ping" ]
+        version_added: '1.3.0'
     monitorid:
         description:
             - ID of the monitor to check.
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:46]: You should document when this is required.

File: plugins/modules/monitoring/uptimerobot.py [hunk 15]
```diff
@@ -8,138 +8,181 @@
...
+        type: int
+        required: false
     apikey:
         description:
             - Uptime Robot API key.
+        type: str
         required: true
 notes:
-    - Support for adding and removing monitors and alert contacts has not yet been implemented.
+    - Support for further functionalities, which the API provides has not yet been implemented.
```
Comment (felixfontein) [inline plugins/modules/monitoring/uptimerobot.py:55]: I'm not sure whether we shouldn'd simply remove this note. It basically says "we didn't implement everything that's theoretically possible", which is the case for almost every module.

File: plugins/modules/monitoring/uptimerobot.py (hunk 4 above)
Comment (felixfontein) [reply plugins/modules/monitoring/uptimerobot.py:60]: This is still missing FQCNs and usage of `name:`. Please don't destroy the work people did to fix all the documentation in this collection.

**Comment by felixfontein** [comment]
@Ninjiner any upates? community.general 1.3.0 will be released in two days, otherwise this has to wait two more months for 2.0.0.

File: plugins/modules/monitoring/uptimerobot.py (hunk 15 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/69567e51172c63f4ab1e7e677337acc391544fe7

File: plugins/modules/monitoring/uptimerobot.py (hunk 4 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Fixed in: https://github.com/ansible-collections/community.general/pull/961/commits/11432b243cbef2ffca238000e905244ec9df39e6

File: plugins/modules/monitoring/uptimerobot.py (hunk 13 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/2b5ae1786b9cd7f4587b452ec7f38452a63d3394

File: plugins/modules/monitoring/uptimerobot.py (hunk 14 above)
Comment (Ninjiner) [reply plugins/modules/monitoring/uptimerobot.py:47]: Changed in: https://github.com/ansible-collections/community.general/pull/961/commits/4f8e5091b2774787ce3871aac8ed1374309d0fa3

**Comment by felixfontein** [comment]
You might have to rebase your PR against the current `main` branch. I guess the 'duplicate `type:`' comes from changes there that aren't visible in this PR.

**Comment by bmillemathias** [comment]
Hi @Ninjiner

Thanks for your PR submission, could you update it ?

Best

File: plugins/modules/monitoring/uptimerobot.py [hunk 16]
```diff
@@ -8,138 +8,180 @@
...
+author:
+- "Nate Kingsley (@nate-kingsley)"
+- "Ninjiner (@Ninjiner)"
 requirements:
     - Valid Uptime Robot API Key
 options:
     state:
         description:
             - Define whether or not the monitor should be running or paused.
+            - C(created), C(absent) and C(present) were added in community.general 1.3.0.
```
Comment (bmillemathias) [inline plugins/modules/monitoring/uptimerobot.py:24]: I would ditch the first sentence and rephrase the second one like "Define the state of the monitor" and explain each state, for the version this is redundant with the field `version_added`.

File: plugins/modules/monitoring/uptimerobot.py [hunk 17]
```diff
@@ -8,138 +8,180 @@
...
+        required: true
+        choices: [ "started", "paused", "created", "absent", "present" ]
```
Comment (bmillemathias) [inline plugins/modules/monitoring/uptimerobot.py:26]: Why is the purpose of the state `created` ?  this is quite unusual to me. 
Should be explained in the `description` field.

**Comment by felixfontein** [comment]
@Ninjiner are you still working on this?

needs_info

**Comment by Ninjiner** [comment]
@felixfontein If I find some time, I am going to update this. Unfortunately, I have spotted some fatal flaws in my initial design of this, which I need to change, which is gonna take me some time. I am thinking about closing this PR for now and open a new one with the rewritten changes.
